File: backend/routers/reports.py
"""
Reports router — sales, orders, waiters, kitchen, bar, date ranges
All reports require manager or admin role
"""
from fastapi import APIRouter, Depends, Query
from typing import Optional
from database import get_db
from routers.auth import require_manager

router = APIRouter()

def fmt_money(v): return round(float(v or 0), 2)

# ── Sales Summary ─────────────────────────────────────────────────────────────
@router.get("/sales/summary")
def sales_summary(
    date_from: Optional[str] = Query(None, description="YYYY-MM-DD"),
    date_to:   Optional[str] = Query(None, description="YYYY-MM-DD"),
    session=Depends(require_manager)
):
    where, params = _date_filter(date_from, date_to, "o.created_at")
    with get_db() as db:
        row = db.execute(f"""
            SELECT
                COUNT(*) as total_orders,
                SUM(CASE WHEN status='paid' THEN 1 ELSE 0 END) as paid_orders,
                SUM(CASE WHEN status='void' THEN 1 ELSE 0 END) as void_orders,
                SUM(CASE WHEN status='paid' THEN subtotal ELSE 0 END) as gross_sales,
                SUM(CASE WHEN status='paid' THEN tax ELSE 0 END) as total_tax,
                SUM(CASE WHEN status='paid' THEN total ELSE 0 END) as net_revenue,
                SUM(CASE WHEN status='paid' AND payment_method='cash'  THEN total ELSE 0 END) as cash_total,
                SUM(CASE WHEN status='paid' AND payment_method='mpesa' THEN total ELSE 0 END) as mpesa_total,
                SUM(CASE WHEN status='paid' AND payment_method='card'  THEN total ELSE 0 END) as card_total,
                COUNT(DISTINCT table_id) as tables_served
            FROM orders o {where}
        """, params).fetchone()
    return {k: fmt_money(v) if isinstance(v, float) else (v or 0)
            for k, v in dict(row).items()}

# ── Sales by Day ──────────────────────────────────────────────────────────────
@router.get("/sales/by-day")
def sales_by_day(
    date_from: Optional[str] = Query(None),
    date_to:   Optional[str] = Query(None),
    session=Depends(require_manager)
):
    where, params = _date_filter(date_from, date_to, "created_at")
    with get_db() as db:
        rows = db.execute(f"""
            SELECT
                DATE(created_at) as day,
                COUNT(*) as orders,
                SUM(CASE WHEN status='paid' THEN total ELSE 0 END) as revenue,
                SUM(CASE WHEN status='paid' THEN tax   ELSE 0 END) as tax
            FROM orders {where}
            GROUP BY DATE(created_at)
            ORDER BY day DESC LIMIT 90
        """, params).fetchall()
    return [dict(r) for r in rows]

# ── Sales by Month ────────────────────────────────────────────────────────────
@router.get("/sales/by-month")
def sales_by_month(year: Optional[int] = None, session=Depends(require_manager)):
    params = []
    where  = ""
    if year:
        where  = "WHERE strftime('%Y', created_at) = ?"
        params = [str(year)]
    with get_db() as db:
        rows = db.execute(f"""
            SELECT
                strftime('%Y-%m', created_at) as month,
                COUNT(*) as orders,
                SUM(CASE WHEN status='paid' THEN total ELSE 0 END) as revenue,
                SUM(CASE WHEN status='paid' THEN tax   ELSE 0 END) as tax
            FROM orders {where}
            GROUP BY strftime('%Y-%m', created_at)
            ORDER BY month DESC LIMIT 24
        """, params).fetchall()
    return [dict(r) for r in rows]

# ── Sales by Year ─────────────────────────────────────────────────────────────
@router.get("/sales/by-year")
def sales_by_year(session=Depends(require_manager)):
    with get_db() as db:
        rows = db.execute("""
            SELECT
                strftime('%Y', created_at) as year,
                COUNT(*) as orders,
                SUM(CASE WHEN status='paid' THEN total ELSE 0 END) as revenue,
                SUM(CASE WHEN status='paid' THEN tax   ELSE 0 END) as tax
            FROM orders
            GROUP BY strftime('%Y', created_at)
            ORDER BY year DESC
        """).fetchall()
    return [dict(r) for r in rows]

# ── Orders Report ─────────────────────────────────────────────────────────────
@router.get("/orders")
def orders_report(
    status:    Optional[str] = Query(None),
    date_from: Optional[str] = Query(None),
    date_to:   Optional[str] = Query(None),
    session=Depends(require_manager)
):
    where, params = _date_filter(date_from, date_to, "o.created_at")
    if status:
        where  += (" AND " if where else "WHERE ") + "o.status=?"
        params.append(status)
    with get_db() as db:
        rows = db.execute(f"""
            SELECT o.*, u.name as waiter_name
            FROM orders o
            LEFT JOIN users u ON o.waiter_id = u.id
            {where}
            ORDER BY o.created_at DESC LIMIT 500
        """, params).fetchall()
        result = []
        for r in rows:
            order = dict(r)
            order["items"] = [dict(i) for i in db.execute(
                "SELECT * FROM order_items WHERE order_id=?", (order["id"],)).fetchall()]
            result.append(order)
    return result

# ── Waiter Performance ────────────────────────────────────────────────────────
@router.get("/waiters")
def waiter_report(
    date_from: Optional[str] = Query(None),
    date_to:   Optional[str] = Query(None),
    session=Depends(require_manager)
):
    where, params = _date_filter(date_from, date_to, "o.created_at")
    with get_db() as db:
        rows = db.execute(f"""
            SELECT
                o.waiter,
                o.waiter_id,
                u.name as user_name,
                COUNT(*) as total_orders,
                SUM(CASE WHEN o.status='paid' THEN 1 ELSE 0 END) as paid_orders,
                SUM(CASE WHEN o.status='void' THEN 1 ELSE 0 END) as void_orders,
                SUM(CASE WHEN o.status='paid' THEN o.total ELSE 0 END) as total_sales,
                AVG(CASE WHEN o.status='paid' THEN o.total END) as avg_order_value
            FROM orders o
            LEFT JOIN users u ON o.waiter_id = u.id
            {where}
            GROUP BY o.waiter_id, o.waiter
            ORDER BY total_sales DESC
        """, params).fetchall()
    return [dict(r) for r in rows]

# ── Kitchen Orders ────────────────────────────────────────────────────────────
@router.get("/kitchen")
def kitchen_report(
    date_from: Optional[str] = Query(None),
    date_to:   Optional[str] = Query(None),
    session=Depends(require_manager)
):
    return _destination_report("kitchen", date_from, date_to)

# ── Bar Orders ────────────────────────────────────────────────────────────────
@router.get("/bar")
def bar_report(
    date_from: Optional[str] = Query(None),
    date_to:   Optional[str] = Query(None),
    session=Depends(require_manager)
):
    return _destination_report("bar", date_from, date_to)

# ── Top Items ─────────────────────────────────────────────────────────────────
@router.get("/top-items")
def top_items(
    destination: Optional[str] = Query(None),
    date_from:   Optional[str] = Query(None),
    date_to:     Optional[str] = Query(None),
    limit: int = 10,
    session=Depends(require_manager)
):
    where, params = _date_filter(date_from, date_to, "o.created_at")
    dest_clause = ""
    if destination:
        dest_clause = (" AND " if where else "WHERE ") + "oi.destination=?"
        params.append(destination)
    with get_db() as db:
        rows = db.execute(f"""
            SELECT
                oi.name,
                oi.destination,
                SUM(oi.qty) as qty_sold,
                SUM(oi.line_total) as revenue
            FROM order_items oi
            JOIN orders o ON oi.order_id = o.id
            {where}{dest_clause}
            AND o.status='paid'
            GROUP BY oi.name, oi.destination
            ORDER BY qty_sold DESC
            LIMIT ?
        """, params + [limit]).fetchall()
    return [dict(r) for r in rows]

# ── Payment Methods ───────────────────────────────────────────────────────────
@router.get("/payments")
def payment_report(
    date_from: Optional[str] = Query(None),
    date_to:   Optional[str] = Query(None),
    session=Depends(require_manager)
):
    where, params = _date_filter(date_from, date_to, "created_at")
    with get_db() as db:
        rows = db.execute(f"""
            SELECT
                payment_method,
                COUNT(*) as count,
                SUM(total) as total
            FROM orders
            {where}
            {"AND" if where else "WHERE"} status='paid'
            AND payment_method IS NOT NULL
            GROUP BY payment_method
        """, params).fetchall()
    return [dict(r) for r in rows]

# ── Void Report ───────────────────────────────────────────────────────────────
@router.get("/voids")
def void_report(
    date_from: Optional[str] = Query(None),
    date_to:   Optional[str] = Query(None),
    session=Depends(require_manager)
):
    where, params = _date_filter(date_from, date_to, "o.created_at")
    with get_db() as db:
        rows = db.execute(f"""
            SELECT o.*, u.name as voided_by_name
            FROM orders o
            LEFT JOIN users u ON o.voided_by = u.id
            {where}
            {"AND" if where else "WHERE"} o.status='void'
            ORDER BY o.created_at DESC LIMIT 200
        """, params).fetchall()
    return [dict(r) for r in rows]

# ── Inventory Valuation ───────────────────────────────────────────────────────
@router.get("/inventory")
def inventory_report(session=Depends(require_manager)):
    with get_db() as db:
        rows = db.execute("""
            SELECT *, (qty_in_stock * cost_per_unit) as stock_value,
            (qty_in_stock <= reorder_level) as low_stock
            FROM inventory ORDER BY category, name
        """).fetchall()
        total = db.execute(
            "SELECT SUM(qty_in_stock * cost_per_unit) as val FROM inventory"
        ).fetchone()["val"] or 0
    return {
        "items": [dict(r) for r in rows],
        "total_stock_value": round(float(total), 2)
    }

# ── Helpers ───────────────────────────────────────────────────────────────────
def _date_filter(date_from, date_to, col):
    clauses, params = [], []
    if date_from:
        clauses.append(f"DATE({col}) >= ?")
        params.append(date_from)
    if date_to:
        clauses.append(f"DATE({col}) <= ?")
        params.append(date_to)
    where = ("WHERE " + " AND ".join(clauses)) if clauses else ""
    return where, params

def _destination_report(dest, date_from, date_to):
    where, params = _date_filter(date_from, date_to, "o.created_at")
    with get_db() as db:
        summary = db.execute(f"""
            SELECT
                COUNT(DISTINCT o.id) as total_orders,
                SUM(oi.qty) as total_items,
                SUM(oi.line_total) as total_revenue
            FROM order_items oi
            JOIN orders o ON oi.order_id = o.id
            {where}
            {"AND" if where else "WHERE"} oi.destination=? AND o.status='paid'
        """, params + [dest]).fetchone()

        items = db.execute(f"""
            SELECT oi.name, SUM(oi.qty) as qty, SUM(oi.line_total) as revenue
            FROM order_items oi JOIN orders o ON oi.order_id=o.id
            {where}
            {"AND" if where else "WHERE"} oi.destination=? AND o.status='paid'
            GROUP BY oi.name ORDER BY qty DESC
        """, params + [dest]).fetchall()
    return {
        "destination": dest,
        "summary": dict(summary),
        "items": [dict(r) for r in items]
    }

File: frontend/static/js/reports.js
/**
 * reports.js — Report rendering for manager/admin
 */
const Reports = (() => {
  const fmt  = n => 'KSh ' + Number(n||0).toLocaleString('en-KE');
  const fmtn = n => Number(n||0).toLocaleString('en-KE');

  async function renderSummary(params = {}) {
    const el = document.getElementById('report-output');
    el.innerHTML = '<div class="empty-state">Loading...</div>';
    try {
      const d = await API.reportSummary(params);
      el.innerHTML = `
        <div class="report-kpi-grid">
          ${kpi('Total Orders',   fmtn(d.total_orders),  '')}
          ${kpi('Paid Orders',    fmtn(d.paid_orders),   'success')}
          ${kpi('Void Orders',    fmtn(d.void_orders),   'danger')}
          ${kpi('Gross Sales',    fmt(d.gross_sales),    'success')}
          ${kpi('VAT Collected',  fmt(d.total_tax),      '')}
          ${kpi('Net Revenue',    fmt(d.net_revenue),    'success')}
          ${kpi('Cash',           fmt(d.cash_total),     '')}
          ${kpi('M-Pesa',         fmt(d.mpesa_total),    '')}
          ${kpi('Card',           fmt(d.card_total),     '')}
          ${kpi('Tables Served',  fmtn(d.tables_served), '')}
        </div>`;
    } catch(e) { el.innerHTML = `<div class="empty-state">Error: ${e.message}</div>`; }
  }

  async function renderByDay(params = {}) {
    const el = document.getElementById('report-output');
    el.innerHTML = '<div class="empty-state">Loading...</div>';
    try {
      const rows = await API.reportByDay(params);
      if (!rows.length) { el.innerHTML = '<div class="empty-state">No data</div>'; return; }
      el.innerHTML = `
        <div class="report-table-wrap">
        <table class="report-table">
          <thead><tr><th>Date</th><th>Orders</th><th>Revenue</th><th>VAT</th></tr></thead>
          <tbody>${rows.map(r=>`<tr>
            <td>${r.day}</td>
            <td>${fmtn(r.orders)}</td>
            <td>${fmt(r.revenue)}</td>
            <td>${fmt(r.tax)}</td>
          </tr>`).join('')}</tbody>
          <tfoot><tr>
            <td><strong>Total</strong></td>
            <td><strong>${fmtn(rows.reduce((s,r)=>s+(r.orders||0),0))}</strong></td>
            <td><strong>${fmt(rows.reduce((s,r)=>s+(r.revenue||0),0))}</strong></td>
            <td><strong>${fmt(rows.reduce((s,r)=>s+(r.tax||0),0))}</strong></td>
          </tr></tfoot>
        </table></div>`;
    } catch(e) { el.innerHTML = `<div class="empty-state">${e.message}</div>`; }
  }

  async function renderByMonth(params = {}) {
    const el = document.getElementById('report-output');
    el.innerHTML = '<div class="empty-state">Loading...</div>';
    try {
      const rows = await API.reportByMonth(params);
      if (!rows.length) { el.innerHTML = '<div class="empty-state">No data</div>'; return; }
      el.innerHTML = `
        <div class="report-table-wrap">
        <table class="report-table">
          <thead><tr><th>Month</th><th>Orders</th><th>Revenue</th><th>VAT</th></tr></thead>
          <tbody>${rows.map(r=>`<tr>
            <td>${r.month}</td>
            <td>${fmtn(r.orders)}</td>
            <td>${fmt(r.revenue)}</td>
            <td>${fmt(r.tax)}</td>
          </tr>`).join('')}</tbody>
        </table></div>`;
    } catch(e) { el.innerHTML = `<div class="empty-state">${e.message}</div>`; }
  }

  async function renderByYear() {
    const el = document.getElementById('report-output');
    try {
      const rows = await API.reportByYear();
      el.innerHTML = `
        <div class="report-table-wrap">
        <table class="report-table">
          <thead><tr><th>Year</th><th>Orders</th><th>Revenue</th><th>VAT</th></tr></thead>
          <tbody>${rows.map(r=>`<tr>
            <td>${r.year}</td>
            <td>${fmtn(r.orders)}</td>
            <td>${fmt(r.revenue)}</td>
            <td>${fmt(r.tax)}</td>
          </tr>`).join('')}</tbody>
        </table></div>`;
    } catch(e) { el.innerHTML = `<div class="empty-state">${e.message}</div>`; }
  }

  async function renderWaiters(params = {}) {
    const el = document.getElementById('report-output');
    el.innerHTML = '<div class="empty-state">Loading...</div>';
    try {
      const rows = await API.reportWaiters(params);
      if (!rows.length) { el.innerHTML = '<div class="empty-state">No data</div>'; return; }
      el.innerHTML = `
        <div class="report-table-wrap">
        <table class="report-table">
          <thead><tr><th>Waiter</th><th>Orders</th><th>Paid</th><th>Voided</th><th>Sales</th><th>Avg Order</th></tr></thead>
          <tbody>${rows.map(r=>`<tr>
            <td><strong>${r.user_name || r.waiter || '—'}</strong></td>
            <td>${fmtn(r.total_orders)}</td>
            <td>${fmtn(r.paid_orders)}</td>
            <td>${fmtn(r.void_orders)}</td>
            <td>${fmt(r.total_sales)}</td>
            <td>${fmt(r.avg_order_value)}</td>
          </tr>`).join('')}</tbody>
        </table></div>`;
    } catch(e) { el.innerHTML = `<div class="empty-state">${e.message}</div>`; }
  }

  async function renderTopItems(params = {}) {
    const el = document.getElementById('report-output');
    el.innerHTML = '<div class="empty-state">Loading...</div>';
    try {
      const rows = await API.reportTopItems(params);
      if (!rows.length) { el.innerHTML = '<div class="empty-state">No data</div>'; return; }
      el.innerHTML = `
        <div class="report-table-wrap">
        <table class="report-table">
          <thead><tr><th>Item</th><th>Destination</th><th>Qty Sold</th><th>Revenue</th></tr></thead>
          <tbody>${rows.map(r=>`<tr>
            <td>${r.name}</td>
            <td><span class="tag-${r.destination}" style="font-size:11px;padding:2px 7px;border-radius:4px">${r.destination}</span></td>
            <td>${fmtn(r.qty_sold)}</td>
            <td>${fmt(r.revenue)}</td>
          </tr>`).join('')}</tbody>
        </table></div>`;
    } catch(e) { el.innerHTML = `<div class="empty-state">${e.message}</div>`; }
  }

  async function renderKitchen(params = {}) { await _renderDest('kitchen', params); }
  async function renderBar(params = {})     { await _renderDest('bar',     params); }

  async function _renderDest(dest, params) {
    const el = document.getElementById('report-output');
    el.innerHTML = '<div class="empty-state">Loading...</div>';
    try {
      const d = await (dest === 'kitchen' ? API.reportKitchen(params) : API.reportBar(params));
      const s = d.summary || {};
      el.innerHTML = `
        <div class="report-kpi-grid" style="margin-bottom:20px">
          ${kpi('Orders', fmtn(s.total_orders), '')}
          ${kpi('Items',  fmtn(s.total_items),  '')}
          ${kpi('Revenue',fmt(s.total_revenue), 'success')}
        </div>
        <div class="report-table-wrap">
        <table class="report-table">
          <thead><tr><th>Item</th><th>Qty</th><th>Revenue</th></tr></thead>
          <tbody>${(d.items||[]).map(r=>`<tr>
            <td>${r.name}</td>
            <td>${fmtn(r.qty)}</td>
            <td>${fmt(r.revenue)}</td>
          </tr>`).join('')}</tbody>
        </table></div>`;
    } catch(e) { el.innerHTML = `<div class="empty-state">${e.message}</div>`; }
  }

  async function renderPayments(params = {}) {
    const el = document.getElementById('report-output');
    try {
      const rows = await API.reportPayments(params);
      el.innerHTML = `
        <div class="report-kpi-grid">
          ${rows.map(r=>kpi(r.payment_method?.toUpperCase()||'Unknown', fmt(r.total)+` (${r.count})`, '')).join('')}
        </div>`;
    } catch(e) { el.innerHTML = `<div class="empty-state">${e.message}</div>`; }
  }

  async function renderVoids(params = {}) {
    const el = document.getElementById('report-output');
    el.innerHTML = '<div class="empty-state">Loading...</div>';
    try {
      const rows = await API.reportVoids(params);
      if (!rows.length) { el.innerHTML = '<div class="empty-state">No void orders in range</div>'; return; }
      el.innerHTML = `
        <div class="report-table-wrap">
        <table class="report-table">
          <thead><tr><th>Order</th><th>Table</th><th>Waiter</th><th>Amount</th><th>Voided By</th><th>Reason</th><th>Date</th></tr></thead>
          <tbody>${rows.map(r=>`<tr>
            <td>${r.id}</td>
            <td>${r.table_id}</td>
            <td>${r.waiter||'—'}</td>
            <td>${fmt(r.total)}</td>
            <td>${r.voided_by_name||'—'}</td>
            <td>${r.void_reason||'—'}</td>
            <td>${(r.created_at||'').slice(0,10)}</td>
          </tr>`).join('')}</tbody>
        </table></div>`;
    } catch(e) { el.innerHTML = `<div class="empty-state">${e.message}</div>`; }
  }

  async function renderInventory() {
    const el = document.getElementById('report-output');
    try {
      const d = await API.reportInventory();
      el.innerHTML = `
        <div class="report-kpi-grid" style="margin-bottom:16px">
          ${kpi('Stock Value', fmt(d.total_stock_value), 'success')}
          ${kpi('Low Stock Items', fmtn(d.items.filter(i=>i.low_stock).length), d.items.some(i=>i.low_stock)?'danger':'')}
        </div>
        <div class="report-table-wrap">
        <table class="report-table">
          <thead><tr><th>Item</th><th>Category</th><th>In Stock</th><th>Unit</th><th>Reorder At</th><th>Cost/Unit</th><th>Value</th><th>Status</th></tr></thead>
          <tbody>${d.items.map(r=>`<tr ${r.low_stock?'style="color:#C8420A"':''}>
            <td>${r.name}</td>
            <td>${r.category}</td>
            <td>${r.qty_in_stock}</td>
            <td>${r.unit}</td>
            <td>${r.reorder_level}</td>
            <td>${fmt(r.cost_per_unit)}</td>
            <td>${fmt(r.stock_value)}</td>
            <td>${r.low_stock?'⚠ Low':'OK'}</td>
          </tr>`).join('')}</tbody>
        </table></div>`;
    } catch(e) { el.innerHTML = `<div class="empty-state">${e.message}</div>`; }
  }

  function kpi(label, value, type='') {
    const colors = {success:'var(--color-background-success)', danger:'var(--color-background-danger)', '':'var(--color-background-secondary)'};
    return `<div class="report-kpi" style="background:${colors[type]||colors['']}">
      <div class="report-kpi-value">${value}</div>
      <div class="report-kpi-label">${label}</div>
    </div>`;
  }

  return { renderSummary, renderByDay, renderByMonth, renderByYear,
           renderWaiters, renderTopItems, renderKitchen, renderBar,
           renderPayments, renderVoids, renderInventory };
})();
